sensor_control/battery_monitor/nodes/battery_monitor.py

```diff
@@ -0,0 +1,162 @@
+
+    def add_to_fr(self, msg):
+        '''
+        Simple method to connect the FR thruster feedback subscriber to the add_voltage method
+        '''
+        self.add_voltage(msg, 1)
+
+    def add_to_bl(self, msg):
+        '''
+        Simple method to connect the BL thruster feedback subscriber to the add_voltage method
+        '''
+        self.add_voltage(msg, 2)
+
+    def add_to_br(self, msg):
+        '''
+        Simple method to connect the BR thruster feedback subscriber to the add_voltage method
+        '''
+        self.add_voltage(msg, 3)
+
+    def add_voltage(self, msg, thruster_id):
+        '''
+        Averages 42 values from each thruster's feedback (42 is a calculated answer to life, the universe, and everything)
+        '''
+        if (self.avg_count[thruster_id] != 42):
+            self.avg_value[thruster_id] += msg.supply_voltage
+            self.avg_count[thruster_id] += 1
+        else:
+            self.supply_voltage[thruster_id] = self.avg_value[thruster_id] / 42
+            self.avg_count[thruster_id] = 0
+            self.avg_value[thruster_id] = 0
+
+    def publish_voltage(self):
+        '''
+        Publishes the average voltage across all four thrustersto the battery_voltage node as a standard Float32 message and runs
+        the voltage_check
+        '''
+        self.voltage = sum(self.supply_voltage) / len(self.supply_voltage)
+        self.pub_voltage.publish(self.voltage)
+        self.voltage_check()
+
+    def voltage_check(self):
+        '''
+        Checks the battery voltage and raises alarms when it's level drops below the set thresholds. This is intended to protect
+        and extend the life of the batteries.
+        '''
+        # A fatal battery warning to kill the boat and protect the batteries
+        if (self.voltage < battery_kill):
+            self.battery_kill_alarm.raise_alarm(
+                problem_description='Bus voltage is at the safety limit; killing the system',
+                parameters={
+                    'bus_voltage': '%f' % (self.voltage),
+                }
+            )
+
+        # A high priority battery warning to abort testing
+        elif (self.voltage < battery_critical):
+            self.battery_critical_alarm.raise_alarm(
+                problem_description='Bus voltage is critical; abort testing',
+                parameters={
+                    'bus_voltage': '%f' % (self.voltage),
+                }
+            )
+
+        # A low priority battery warning to inform user's of the status
+        elif (self.voltage < battery_low):
+            self.battery_low_alarm.raise_alarm(
+                problem_description='Bus voltage is approaching safety limit',
+                parameters={
+                    'bus_voltage': '%f' % (self.voltage),
+                }
+            )
+
+
+if __name__ == "__main__":
+    monitor = BatteryMonitor()
+    rate = rospy.Rate(1)
+    while not rospy.is_shutdown():
```

Comment:
http://wiki.ros.org/rospy/Overview/Time
Check out Timers, they replace this whole thing with one line